Griffin ABTR XS generation and steady-state analysis workflow (#387)
idaholab/virtual_test_bed: doc/content/sfr/abtr_xsgen_workflow/abtr_xsgen_workflow.md

- Homogenized volume fractions: For each subassembly ID shown in [tab:subassembly_id_mapping], a list of volume fractions for each heterogeneous region ID found within the subassembly is stored. This is used in the Step 1 MC$^2$-3 calculation step to set the compositions of the homogenized mixture calculation for each subassembly region.
- 1-D radial boundaries: For each subassembly ID shown in [tab:subassembly_id_mapping], the heterogeneous subassembly geometry and region ID layout is represted as a series of volume-preserved concentric rings, and this 1-D ring representation is used by the Step 2 MC$^2$-3 calculation to setup the 1-D CPM calculations wherever needed.

For a more comprehensive list of metadata stored by `EquivalentCoreMeshGenerator`, the console output from setting `EquivalentCoreMeshGenerator/debug_equivalent_core = true` can be inspected. As a final step for mesh generation, an explicity defined coarse mesh is defined for the output R-Z mesh in order to use with [!ac](CMFD) acceleration as part of the downstream Griffin R-Z spectrum calculation step, and setting `Mesh/coord_type = RZ` specifies that an RZ coordinate system should be used for this.

> !listing sfr/abtr_xsgen_workflow/xs_generation_mesh.i start=Define and apply coarse mesh to RZ mesh

Author: By original heterogeneous RZ mesh, do you mean an RZ mesh that represents pin-level heterogeneity? If so, then no. The RZ mesh is based on an assembly homogenized lattice configuration

> !listing sfr/abtr_xsgen_workflow/xs_generation_mesh.i start=Define and apply coarse mesh to RZ mesh

Reviewer: ok then I like the amended better. If you want more detail that works too

```
[Mesh]
  [coarse_mesh]
    type = CoarseMeshExtraElementIDGenerator
    input = block_numbering
    coarse_mesh = block_numbering
    extra_element_id_name = coarse_element_id
  []  
[]
```

Here, the coarse mesh is defined to be identical to the fine mesh, which in this case is the assembly-homogeneous mesh created by `EquivalentCoreMeshGenerator`. Thus, the full-core eigenvalue calculation can be run by running the following commands:

```
# Run full-core eigenvalue calculation and override default parameters defined in core_hom_macro.i
mpirun -np <N_PROC_TOTAL> /path/to/griffin/griffin-opt -i abtr_het_mesh.i core_hom_macro.i
    Mesh/eqv_core/quad_center_elements=true Mesh/eqv_core/max_axial_mesh_size=10 Mesh/uniform_refine=1 
    TransportSystems/ReflectingBoundary='' TransportSystems/VacuumBoundary='outer_core top bottom'
    TransportSystems/dfem_sn/NAzmthl=4 TransportSystems/dfem_sn/NPolar=3
    Executioner/coarse_element_id=coarse_element_id Executioner/cmfd_eigen_solver_type=krylovshur
    PowerDensity/power=250000000 PowerDensity/power_density_variable=power --distributed-mesh
```

> Here, mesh refinement is applied and `EquivalentCoreMeshGenerator` settings are updated through the `[Mesh]` block, the boundary conditions sidesets are set and DFEM-SN angular discretizations are updated to increase solution accuracy through the `[TransportSystems]` block, CMFD settings are updated through the `[Executioner]` block, and the reactor power is set through the `[PowerDensity]` block. These parameter settings ensure that calculated eigenvalues are predicted to within 100pcm of previous work that has looked at modeling the [!ac](ABTR) reactor core [!citep](pyarc_report).

Reviewer: a list of the changes would be more readable?
```suggestion
The command line options specify that:

 - mesh refinement is applied
 - `EquivalentCoreMeshGenerator` settings are updated through the `[Mesh]` block
 - the boundary conditions sidesets are set
 - DFEM-SN angular discretizations are updated to increase solution accuracy through the `[TransportSystems]` block
 - CMFD settings are updated through the `[Executioner]` block
 - the reactor power is set through the `[PowerDensity]` block.
 
These parameter settings ensure that calculated eigenvalues are predicted to within 100pcm of previous modeling work for the [!ac](ABTR) reactor core [!citep](pyarc_report).
```

The cross section generation workflow has been implemented in such a way that many of the input file generation steps, data transfers, and generation of input meshes and problem boundaries related to how Griffin and MCC-3 operate have been automated. It should also be noted that all intermediate meshes and composition mappings are also automatically inferred from the input heterogeneous mesh, which directly contains all region ID and mesh specifications for each pin, assembly, and core region. Users who require more background on this cross section generation workflow are referred to [!citep](MCC3manual) and [!citep](lee2023improved). The following section details how to define the heterogeneous input mesh through the use of RGMB mesh generators, as well as how to set the control parameters within the Griffin input file to dictate how MCC-3 and Griffin are executed within the three steps that comprise the cross section generation workflow.

## Input File Description for Cross Section Generation Workflow id=sec:xsgen_input_description

> This section focuses on the input files needed to run the cross section generation worfklow. Two input files are called for this section in order to split mesh generation from the actual execution of the cross section generation workflow calculations. `xs_generation_mesh.i` defines the input heterogeneous mesh through [!ac](RGMB) mesh generators and generates the equivalent R-Z mesh that will used as part of the Griffin R-Z spectrum calculation, while `xs_generation.i` imports this mesh through [FileMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/FileMeshGenerator.html) and defines all control parameters related to the cross section generation workflow. The commands to run the cross section generation workflow from start to finish are shown below:

Reviewer: ```suggestion
This section focuses on the input files needed to run the cross section generation worfklow. Two input files are called for this section in order to split mesh generation from the actual execution of the cross section generation workflow calculations. `xs_generation_mesh.i` defines the input heterogeneous mesh through [!ac](RGMB) mesh generators and generates the equivalent R-Z mesh that will be used as part of the Griffin R-Z spectrum calculation, while `xs_generation.i` imports this mesh through [FileMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/FileMeshGenerator.html) and defines all control parameters related to the cross section generation workflow. The commands to run the cross section generation workflow from start to finish are shown below:
```

```
# Generate equivalent R-Z spectrum mesh and run XS generation workflow through Griffin
mpirun -np [N_PROC_TOTAL] /path/to/griffin/griffin-opt -i xs_generation_mesh.i xs_generation.i
```

, where `[N_PROC_TOTAL]` refers to the total number of MPI processes available for the simulation. The first Griffin file `xs_generation_mesh.i` generates an output 2-D R-Z mesh from an input mesh that represents the geometry and material ID assignment for the 3-D heterogeneous [!ac](ABTR) core. The second Griffin input file `xs_generation.i` contains all of the information related to the cross section generation worklow. [#subsec:rgmb_meshgen] and [#subsec:rz_eqv_core] describe sections of `xs_generation_mesh.i` in more detail, while [#subsec:mcc3_params] and [#subsec:griffin_rz_params] describe the specifics of `xs_generation.i`.

### Heterogeneous input mesh generation through [!ac](RGMB) mesh generators id=subsec:rgmb_meshgen

> The use of [ReactorMeshParams](https://mooseframework.inl.gov/source/meshgenerators/ReactorMeshParams.html), [PinMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/ReactorMeshParams.html), [AssemblyMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/AssemblyMeshGenerator.html), and [CoreMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/CoreMeshGenerator.html) in `xs_generation_mesh.i` make use of the [!ac](RGMB) mesh generation system within the Reactor module to streamline the definition of latticed core geometries and the reporting IDs associated with such structures [!citep](shemon2023reactor). Mesh generation through [!ac](RGMB) is a requirement for taking advantage of Griffin's [!ac](SFR) cross section generation workflow, as it ensures that all reporting IDs related to region ID mappings as well as IDs corresponding to groupings of pin structures, assembly structures, and axial layers are defined in a consistent and predictable manner.

Reviewer: ```suggestion
The [ReactorMeshParams](https://mooseframework.inl.gov/source/meshgenerators/ReactorMeshParams.html), [PinMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/ReactorMeshParams.html), [AssemblyMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/AssemblyMeshGenerator.html), and [CoreMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/CoreMeshGenerator.html) in `xs_generation_mesh.i` make use of the [!ac](RGMB) mesh generation system within the Reactor module to streamline the definition of latticed core geometries and the reporting IDs associated with such structures [!citep](shemon2023reactor). Mesh generation through [!ac](RGMB) is a requirement for taking advantage of Griffin's [!ac](SFR) cross section generation workflow, as it ensures that all reporting IDs related to region ID mappings as well as IDs corresponding to groupings of pin structures, assembly structures, and axial layers are defined in a consistent and predictable manner.
```

Additionaly, setting `MCC3CrossSection/rz_calculation` to true indicates that an intermediate Griffin flux RZ calculation should be conducted as part of the cross section generation workflow. `MCC3CrossSection/rz_spectrum_pn_order` sets the maximum $P_N$ flux moment order of the RZ flux spectrum, `MCC3CrossSection/rz_group_structure` sets the group structure of the intermediate flux spectrum calculation, while `MCC3CrossSection/group_structure` sets the final group structure of the output cross section library after running the Step 2 MC$^2$-3 calculation.

> `MCC3CrossSection/map_het_grid_name`, `MCC3CrossSection/map_het_grid_ref_value`, and `MCC3CrossSection/map_het_grid_values` control how cross sections are tabulated as a function of tabulation variables such as temperature. `MCC3CrossSection/map_het_grid_name` maps each compisition defined in the `[Compositions]` block that has an associated region ID to an associated grid variable. For the purposes of this tutorial, each heterogeneous region has a single temperature variable it is tabulated against, where `Tfuel` represents the fuel temperature, and `Tcool` represents the coolant temperature. `MCC3CrossSection/map_het_grid_ref_value` sets the reference value for each grid variable of interest, while `MCC3CrossSection/map_het_grid_values` defines all tabulated values that are assocaited with each grid variable.

Reviewer: ```suggestion
`MCC3CrossSection/map_het_grid_name`, `MCC3CrossSection/map_het_grid_ref_value`, and `MCC3CrossSection/map_het_grid_values` control how cross sections are tabulated as a function of tabulation variables such as temperature. `MCC3CrossSection/map_het_grid_name` maps each composition defined in the `[Compositions]` block that has an associated region ID to an associated grid variable. For the purposes of this tutorial, each heterogeneous region has a single temperature variable it is tabulated against, where `Tfuel` represents the fuel temperature, and `Tcool` represents the coolant temperature. `MCC3CrossSection/map_het_grid_ref_value` sets the reference value for each grid variable of interest, while `MCC3CrossSection/map_het_grid_values` defines all tabulated values that are assocaited with each grid variable.
```

## Input File Description for Full-Core Eigenvalue Calculation id=sec:fullcorecalc_input_description

By running Griffin on the `xs_generation.i` file, two files are created that are used for the full-core eigenvalue calculation step that will be explained in this section. The first is the BG cross section library, named `mcc3xs.xml` (this name is controlled by `MCCCrossSection/xml_filename`), and the second is the Griffin input file that will be used for this section, named `core_hom_macro.i` (this is a fixed name, based on whether assembly homogenized cross sections are calculated, and whether microscopic or macroscopic cross sections are used in the output cross section library). The automatic generation of this input file streamlines the setup of the full-core analysis step, and the contents of this file are shown below:

> 

Reviewer: use a listing this wont age well

    type = MixedMatIDNeutronicsMaterial
    isotopes = 'pseudo_REG'
    densities = '1.0'
    block = '100 300 400 700 800 900 1000'
    grid_names = 'Tcool'
    grid = '1'
  []
  [material_core_grid2]
    type = MixedMatIDNeutronicsMaterial
    isotopes = 'pseudo_REG'
    densities = '1.0'
    block = '200 500 600'
    grid_names = 'Tcool Tfuel'
    grid = '1 1'
  []
[]
```

Users are encouraged to look through this input file in detail and update input parameters accordingly based on their modeling needs. At a minimum however, a mesh file related to the heterogeneous reactor core needs to be defined with a name that matches `Mesh/eqv_core/input`, and boundary conditions for the full-core eigenvalue problem need to be defined by setting the appropriate outer boundary sidesets using `TransportSystems/VacuumBoundary` and `TransportSystems/ReflectingBoundary`. By looking at this file, it can be seen that the mapping between mesh blocks and cross section material ID is automatically done by calling `MixedMatIDNeutronicsMaterial` in the `[Materials]` block, and this assumes that the homogeneous input reactor core mesh has been generated with `EquivalentCoreMeshGenerator`.

> The heterogeneous reactor core mesh can be copied directly from the `xs_generation_mesh.i` (everything up to and including the `Mesh/het_core` block), and pasted into `core_hom_macro.i` above the `Mesh/eqv_core` block. The remainder of the `[Mesh]` block in `core_hom_macro.i` follows very similar steps as the ones explained in `xs_generation_mesh.i`, `EquivalentCoreMeshGenerator` is responsible for generating the homogeneous mesh from the input heterogeneous mesh specifications.

Reviewer: I wonder if you could
!include that input file instead, with the right 'active/inactive' parameter in [Mesh]
or maybe with a final_generator_name

Author: @GiudGiud in case it wasn't clear through the documentation, Griffin will automatically generate a `core_hom_macro.i` file that will be the starting point for running the full core solve step. The user will have to modify this file slightly to get it to actually run for this problem, so I'm wondering if it makes sense to just commit a file `core_hom_macro_updated.i` that has all the additions needs for the full core solve step to properly run. Previously I was trying to do this all through cli args in order to make sure that the process of file generation and file editing could be tested properly, but the downside of this is that the documentation page does not have a file `core_hom_macro.i` to reference. What do you suggest?

Author: The other way this could potentially be simplified is to define another file `abtr_het_mesh.i` which contains the shared mesh blocks between `xs_generation_mesh.i` and `core_hom_macro.i`, since the het mesh generation step is exactly the same in both cases.

Author: Basically, the slightly confusing thing here is explaining what changes are needed to `core_hom_macro.i` to define the Mesh block. Mesh blocks needed to be added to both the beginning (het mesh definition) and the end of the mesh block (coarse mesh for equivalent core) in `core_hom_macro.i`

Author: @GiudGiud I've gone ahead and added the file `abtr_het_mesh.i`, which has all the meshing routines that are common to both xs generation and full core solve steps. The only block in the documentation that is manually defined is the `[coarse_mesh]` block, since this is the only block that is specific to the full core solve step. Let me know if this is sufficient

In `core_hom_macro.i`, `EquivalentCoreMeshGenerator/target_geometry` is set to `full_hom` to specify a full-homogeneous reactor mesh as the output mesh. Additionally, `EquivalentCoreMeshGenerator/quad_center_elements` can be set to dictate whether the homogeneized hexagonal prism regions are discretized into 2 quadrilaterial prism regions (true) or 6 triangular prism regions (false), and `EquivalentCoreMeshGenerator/max_axial_mesh_size` sets the maximum axial mesh size to sub-discretize each axial layer of the output mesh.

!media plots/eqv_fullhomcore.png
       style=width:80%
       id=fig:eqv_fullhomcore
       caption=Generation of 3-D fully homogeneous core mesh (right) from 3-D heterogeneous core created through [!ac](RGMB) mesh generators (left).

While `core_hom_macro.i` does not define a coarse mesh for the output mesh, this can be done manually by adding the following lines to the end of the `[Mesh]` block in `core_hom_macro.i`:

```

> [Mesh]

Reviewer: same

Author: @GiudGiud this is not part of any input file listing right now, since core_hom_macro.i does not define a coarse mesh. Should I add these parameters in the CLI section where I discuss how to run the problem?
